### PR TITLE
Add LIFO functions to 'ring-buffer type.

### PR DESCRIPTION
**Pull request by noogie13**
Accidentally closed last PR.

Split into three commits.

### CODE REVIEW

**Comment by noogie13** [comment]
Do we need to add some `next-item` things for the LIFO implementation? It doesn't seem to have any obvious uses, maybe I just don't notice?

**Comment by Ambrevar** [comment]
Finally found time to look at this, sorry for the delay!

Thinking a bit more about it, I think "last-in-first-out" should be an container slot and all methods should change their behaviour depending on the value of this slot.

Rationale: the container is meant to be generic, i.e. to use the same function as any other container.  Say, you have a list of containers (lists, ring-buffers, vectors) in a list and you want index X in all those containers: the generic `item-at` will pick the wrong index in a LIFO ring-buffer.

Does that make sense?  Any other idea?

Suggestions:

- `delete-item` should be named `delete-item-at`.

- Can you implement `delete-item` too?  It takes an item as argument, see the others uses in the library.

- To implement the delete-* defmethods, I believe ring-buffer should be an ordered-container-mixin.

- Once you've adapted `item-at`, `first-element`, etc. to obbey the last-in-first-out slot, you can have all other functions depend on those methods.  For instance, at the moment `iterate-nodes` uses `buffer-start` and `buffer-end` which is wrong for LIFOs, so it should be changed to use `first-element` and `(item-at (size ...))`.

Thoughts?

Let me know if you need help with this, I can also work on it.

**Comment by noogie13** [comment]
It seems the `item-at` function is just reducing the ring-buffer to the list that's underneath it, ignoring all the existing FIFO logic. Am I misreading? What should I change?

**Comment by Ambrevar** [comment]
`item-at` can take multiple indices in other containers, but apparently for ring-buffers it was decided to ignore the indices after the first one, hence the `(car indexes)`.

What you need to do here is check the value of last-in-first-out-flag: if true, return the lifo-indexed element, otherwise return the value originally returned by `item-at`.


In `make-ring-buffer` I suggest you make `last-in-first-out` optional and default to nil.

**Comment by noogie13** [comment]
Do you think it's better etiquette to define a ring-buffer and write LIFO and FIFO subclasses?

**Comment by Ambrevar** [comment]
It's a good idea.  What about the following:

- The current ring-buffer class is the FIFO one.

- Overhaul the methods so that it's easy to change the direction.

- Create a new ring-buffer-reverse that inherits from ring-buffer and
  overloads the few methods where the direction matters.

What do you think?

**Comment by noogie13** [comment]
Seems good. Maybe reverse should be the default, but I guess we can leave it how it is.

Also will have to adapt our deleting functions and such. I'll start working on it.

**Comment by Ambrevar** [comment]
Great!

We can't change the default because that would break backward
compatibility :(

**Comment by noogie13** [comment]
Shall I extend our `item-at`'s to ref multiple indices?

**Comment by Ambrevar** [comment]
If other containers support this, it makes sense.  Would be nice to know
why it wasn't supported in the first place.  Maybe out of laziness?

**Comment by noogie13** [comment]
Another question: `delete-item-at` for ring-buffer-reverse is our LIFO one that keeps the LIFO-indexes intact, but `delete-item-at` for ring-buffer, should I just NIL the item? What do you think?

**Comment by noogie13** [comment]
Also, still don't understand this `current-item` nonsense. Doesn't seem useful to me, again maybe I'm not seeing it's usefulness?

**Comment by Ambrevar** [comment]
delete-item-at: Why just NIL the item?  To me it seems that you need to shift all the elements, just like for ring-buffer-reverse.  Actually, beside the index, it's the same function for both ring-buffers if I'm not mistaken.

current-item: indeed, it does not seem to make sense.  I think there was some confusion by the time the library was written precisely between FIFO and LIFO.  To keep backward compatibility, I suggest we make `current-item` return the last index.  For ring-buffer, that's the buffer end (as currently implemented), for ring-buffer-reverse, that's the buffer start.

File: dev/ring-buffers.lisp [hunk 1]
```diff
@@ -24,11 +21,12 @@
    (total-size :reader total-size
                :initarg :total-size)))
 
+(defclass* ring-buffer-reverse (ring-buffer) ())
+
 (defun make-ring-buffer (size last-in-first-out)
```
Comment (Ambrevar) [inline dev/ring-buffers.lisp]: I think it'd be more convenient with `(size &optional last-in-first-out)`.

**Comment by noogie13** [comment]
`current-item` for ring-buffer-reverse should be the buffer start? Why not just lifo-index 0 (most recent added item). 

Also I'm not sure how to implement delete-item-at for ring-buffer, because I'm not sure why one would want to delete from ring-buffer by arbitrary underlying vector index. 

There should really be no difference for a FIFO and LIFO data structure except the dequeue behavior (irrelevant here), not sure what this ring-buffer with raw vector indexing has to do with it. If we wanted our data structure to be FIFO we could just write a `fifo-index` function similar to the `lifo-index` that I already implemented.

**Comment by Ambrevar** [comment]
Solomon Bloch <notifications@github.com> writes:

> `current-item` for ring-buffer-reverse should be the buffer start? Why not just lifo-index 0 (most recent added item).

Why not.  Then both LIFO and FIFO's current-item would point to the same
item, right?
Probably no big deal.


> Also I'm not sure how to implement delete-item-at for ring-buffer,
> because I'm not sure why one would want to delete from ring-buffer by
> arbitrary underlying vector index.

Simple use case: imagine the user wants to clear specific items in a
history.
Emacs can do this for instance.

> There should really be no difference for a FIFO and LIFO data
> structure except the dequeue behavior (irrelevant here), not sure what
> this ring-buffer with raw vector indexing has to do with it. If we
> wanted our data structure to be FIFO we could just write a
> `fifo-index` function similar to the `lifo-index` that I already
> implemented.

I'm not sure what you are asking here.
Indeed, the FIFO and LIFO structures should be the same except for one
method, the one that retrieves the element at a given index.
If all other methods depend on this index method, then we are done.

The problem is that at the moment, many methods touch the underlying
vector manually instead of using the index method.  We need to fix this.

Makes sense?

**Comment by Ambrevar** [comment]
Friendly ping! ;)

**Comment by noogie13** [comment]
Working on this again; was busy with finals at university!

**Comment by Ambrevar** [comment]
Great, thanks!

**Comment by Ambrevar** [comment]
Hi again!  Did you find time to make progress on this?  Cheers!

**Comment by noogie13** [comment]
I finished awhile ago, but should probably test to make sure it all works properly.

File: dev/ring-buffers.lisp [hunk 2]
```diff
@@ -9,48 +9,119 @@
...
+
+(defmethod item-at ((container ring-buffer-reverse) &rest indexes)
+  (declare (dynamic-extent indexes))
+  (let ((indexes (mapcar #'(lambda (index) (lifo-index container index))
+                         indexes)))
+    (svref (contents container)
+           (mod (first indexes) (total-size container)))))
 
-(defmethod item-at! ((container ring-buffer) value &rest indexes)
+(defmethod item-at! ((container ring-buffer-reverse) value &rest indexes)
```
Comment (Ambrevar) [inline dev/ring-buffers.lisp]: I think the specialization should be against ring-buffer, not ring-buffer-reverse here (it's already there below).  Anyways, we might not need 2 item-at* functions, see below.

File: dev/ring-buffers.lisp [hunk 3]
```diff
@@ -9,48 +9,119 @@
...
+(defmethod lifo-index ((container ring-buffer-reverse) index)
+  "Return index converted to internal LIFO index, where items are ordered from newest to oldest."
+  (mod (1- (+ (buffer-start container)
+              (- (buffer-end container) (buffer-start container) index)))
+       (total-size container)))
+
+(defmethod fifo-index ((container ring-buffer) index)
+  "Return index converted to internal FIFO index, where items are ordered from oldest to newest."
+  (mod (+ (buffer-start container) index)
+       (total-size container)))
```
Comment (Ambrevar) [inline dev/ring-buffers.lisp]: Another suggestion: rename both `lifo-index` and `fifo-index` to the same name (e.g. `index` if that's available).  Then you can remove the `item-at*` ring-buffer-reverse specialization.  Makes sense?

File: dev/ring-buffers.lisp [hunk 4]
```diff
@@ -9,48 +9,119 @@
...
+      (incf buffer-start))
+    (incf buffer-end)))
+
+(defmethod delete-item-at ((container ring-buffer) &rest indexes)
+  "Delete item using FIFO index for ring-buffer."
+  (delete-item-internal container (mapcar #'(lambda (x) (fifo-index container x)) indexes)))
+
+(defmethod delete-item-at ((container ring-buffer-reverse) &rest indexes)
+  "Delete item using LIFO index for ring-buffer-reverse.."
+  (delete-item-internal container (mapcar #'(lambda (x) (lifo-index container x)) indexes)))
```
Comment (Ambrevar) [inline dev/ring-buffers.lisp]: Same here: with a specialized `index` method you don't need to specialize `delete-item-at`.

File: dev/ring-buffers.lisp [hunk 5]
```diff
@@ -9,48 +9,119 @@
...
+
+(defmethod delete-item-internal ((container ring-buffer) &rest indexes)
+  "Delete item using internal index (not meant to be used directly)."
```
Comment (Ambrevar) [inline dev/ring-buffers.lisp]: Prefix the name with `%` to mark it as "internal".  No need to mention "internal" then.

File: dev/ring-buffers.lisp [hunk 6]
```diff
@@ -9,48 +9,119 @@
...
+  (let ((index (car indexes)))
+    (with-slots (buffer-end buffer-start contents total-size) container
+      (cond ((= (mod index total-size) (1- buffer-end))
```
Comment (Ambrevar) [inline dev/ring-buffers.lisp]: Maybe insert a new line after `cond` to reduce indentation.

File: dev/ring-buffers.lisp [hunk 7]
```diff
@@ -9,48 +9,119 @@
...
-     (remf args :total-size)
-     (make-ring-buffer total-size)))
+  (let ((total-size (getf args :total-size 1))
+        (last-in-first-out (getf args :last-in-first-out nil)))
+    (remf args :total-size)
+    (remf args :last-in-first-out)
+    (make-ring-buffer total-size last-in-first-out)))
 
 
 ;;?? the (first indexes) is odd...
```
Comment (Ambrevar) [inline dev/ring-buffers.lisp]: Maybe document that only the first index is used.

File: dev/ring-buffers.lisp [hunk 8]
```diff
@@ -9,48 +9,119 @@
...
+  "Delete item using FIFO index for ring-buffer."
+  (delete-item-internal container (mapcar #'(lambda (x) (fifo-index container x)) indexes)))
+
+(defmethod delete-item-at ((container ring-buffer-reverse) &rest indexes)
+  "Delete item using LIFO index for ring-buffer-reverse.."
+  (delete-item-internal container (mapcar #'(lambda (x) (lifo-index container x)) indexes)))
+
+(defmethod delete-item-internal ((container ring-buffer) &rest indexes)
+  "Delete item using internal index (not meant to be used directly)."
+  (let ((index (car indexes)))
```
Comment (Ambrevar) [inline dev/ring-buffers.lisp]: Why deleting only the first index?  We could just as well loop over all indices, it would be less confusing with regard to the method arguments.

**Comment by Ambrevar** [comment]
Great!  Looks good, almost ready to be merged.  My main pick is that we can generalize the `fifo-index` and `lifo-index` functions and remove the other duplicate methods.  What do you think?

**Comment by Ambrevar** [comment]
It would be great to add tests, but since there were none before and your changes don't break backward compatibility, it can always be added later.

File: dev/ring-buffers.lisp (hunk 2 above)
Comment (noogie13) [reply dev/ring-buffers.lisp]: What do you mean the specialization against ring-buffer? Also, yes, we don't need two item-at* functions. I can't believe I wrote multiple versions of the same method only differing by the index function. Obviously poor quality, oops.

File: dev/ring-buffers.lisp (hunk 3 above)
Comment (noogie13) [reply dev/ring-buffers.lisp]: Yes... I should've done this from the jump. Oops.

File: dev/ring-buffers.lisp (hunk 8 above)
Comment (noogie13) [reply dev/ring-buffers.lisp]: Ah, good point. I did this to conform with what was preexisting, vis-a-vis a list of inputs. I will change that.

File: dev/ring-buffers.lisp (hunk 2 above)
Comment (Ambrevar) [reply dev/ring-buffers.lisp]: Both `item-at!` methods specialized on ring-buffer-reverse.  One should have specialized on `ring-buffer`.

**Comment by Ambrevar** [comment]
I've pushed your changes after squashing them in f4fb24a4be6737179afa7e6c4fc3ed4e70a38a07.

I've fixed a couple of things afterwards:
- %index now wraps over the `size`, not the `total-size` to avoid indexing elements outside of the buffer-start - buffer-end range.
- I've simplified delete-item-at which now accepts only one index, like item-at.

**Comment by Ambrevar** [comment]
Thanks for your patience, we've finally made it! :)
